src/LobbyTypes.hs

```diff
@@ -34,10 +34,50 @@ type GamesList = CC.MVar [LobbyGame]
 type LobbyState = (Server ConcurrentClientList, Server GamesList, Server ConcurrentChatList)
 
 -- |A chat message sent on a channel.
-data ChatMessage = ChatMessage {from    :: SessionID
-                               ,content :: String}
+data ChatMessage = ChatMessage {from    :: Name
+                              ,content :: String}
+                | ChatJoin
+                | ChatAnnounceJoin {from :: Name}
+                | ChatLeave
+                | ChatAnnounceLeave {from :: Name}
+
 -- |A chat has a name and all sessionIDs currently in the chat.
-type Chat = (Name,[SessionID])
+type Chat = (Name, CC.Chan ChatMessage)
+
+instance Binary ChatMessage where
+  put (ChatMessage from content) = do
+    put (0 :: Word8)
+    put from
+    put content
+  put (ChatJoin) =
+    put (1 :: Word8)
+  put (ChatAnnounceJoin from) = do
+    put (2 :: Word8)
+    put from
+  put (ChatLeave) =
+    put (3 :: Word8)
+  put (ChatAnnounceLeave from) = do
+    put (4 :: Word8)
+    put from
```

Comment:
Maybe have a function for each put, looks somewhat confusing right now? Something like:
```
put (ChatMessage from content) = putChatMessage from content
put ChatJoin                   = putChatJoin
put (ChatAnnounceJoin from)    = putChatAnnounceJoin from
etc...
  where
    putChatMessage from content = do
       put (0 :: Word8)
       put from
       put content
  putChatJoin...
```
Just to increase readability?